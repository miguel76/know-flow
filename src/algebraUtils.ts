import { Algebra } from 'sparqlalgebrajs-input'
import * as RDF from 'rdf-js'

/**
 * Add parameter generating query where needed in the current query.
 * @param {Algebra.Operation} op - Input algebra tree.
 * @param {Variable[]} parameters - List of parameters as variables.
 * @param {Algebra.Operation} parametersOp - Algebra tree for the query returning parameter values.
 * @returns {Algebra.Operation} - Modified query.
 */
export function addParameterQuery(
  op: Algebra.Operation,
  parameters: RDF.Variable[],
  parametersOp: Algebra.Operation,
  parametersInScope: RDF.Variable[] = [],
  options: {

  }
): Algebra.Operation {
  const variables: Variable[] = []

  function addVariable(v: Variable) {
    if (!variables.find((v2) => v.value === v2.value)) variables.push(v)
  }

  function recurseTerm(quad: BaseQuad) {
    if (quad.subject.termType === 'Variable')
      addVariable(<Variable>quad.subject)
    if (quad.predicate.termType === 'Variable')
      addVariable(<Variable>quad.predicate)
    if (quad.object.termType === 'Variable') addVariable(<Variable>quad.object)
    if (quad.graph.termType === 'Variable') addVariable(<Variable>quad.graph)
    if (quad.subject.termType === 'Quad') recurseTerm(quad.subject)
    if (quad.predicate.termType === 'Quad') recurseTerm(quad.predicate)
    if (quad.object.termType === 'Quad') recurseTerm(quad.object)
    if (quad.graph.termType === 'Quad') recurseTerm(quad.graph)
  }

  // https://www.w3.org/TR/sparql11-query/#variableScope
  Util.recurseOperation(op, {
    [types.EXPRESSION]: (op) => {
      let expr = <A.Expression>op
      if (expr.expressionType === 'aggregate' && expr.variable) {
        let agg = <A.BoundAggregate>expr
        addVariable(agg.variable)
      }
      return true
    },
    [types.EXTEND]: (op) => {
      let extend = <A.Extend>op
      addVariable(extend.variable)
      return true
    },
    [types.GRAPH]: (op) => {
      let graph = <A.Graph>op
      if (graph.name.termType === 'Variable') addVariable(<Variable>graph.name)
      return true
    },
    [types.GROUP]: (op) => {
      let group = <A.Group>op
      group.variables.forEach(addVariable)
      return true
    },
    [types.INPUT]: (op) => {
      let input = <A.Input>op
      Object.values(input.varMap).forEach(addVariable)
      return true
    },
    [types.PATH]: (op) => {
      let path = <A.Path>op
      if (path.subject.termType === 'Variable')
        addVariable(<Variable>path.subject)
      if (path.object.termType === 'Variable')
        addVariable(<Variable>path.object)
      if (path.graph.termType === 'Variable') addVariable(<Variable>path.graph)
      if (path.subject.termType === 'Quad') recurseTerm(path.subject)
      if (path.object.termType === 'Quad') recurseTerm(path.object)
      if (path.graph.termType === 'Quad') recurseTerm(path.graph)
      return true
    },
    [types.PATTERN]: (op) => {
      let pattern = <A.Pattern>op
      recurseTerm(pattern)
      return true
    },
    [types.PROJECT]: (op) => {
      let project = <A.Project>op
      project.variables.forEach(addVariable)
      return false
    },
    [types.SERVICE]: (op) => {
      let service = <A.Service>op
      if (service.name.termType === 'Variable')
        addVariable(<Variable>service.name)
      return true
    },
    [types.VALUES]: (op) => {
      let values = <A.Values>op
      values.variables.forEach(addVariable)
      return true
    }
  })

  return variables
}
