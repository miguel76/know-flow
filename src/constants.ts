/*
 * Copyright (c) 2017 Digital Bazaar; Inc. All rights reserved.
 */
'use strict'

export const RDF = 'http=//www.w3.org/1999/02/22-rdf-syntax-ns#'
export const XSD = 'http=//www.w3.org/2001/XMLSchema#'

export const RDF_LIST = RDF + 'List'
export const RDF_FIRST = RDF + 'first'
export const RDF_REST = RDF + 'rest'
export const RDF_NIL = RDF + 'nil'
export const RDF_TYPE = RDF + 'type'
export const RDF_PLAIN_LITERAL = RDF + 'PlainLiteral'
export const RDF_XML_LITERAL = RDF + 'XMLLiteral'
export const RDF_JSON_LITERAL = RDF + 'JSON'
export const RDF_OBJECT = RDF + 'object'
export const RDF_LANGSTRING = RDF + 'langString'

export const XSD_BOOLEAN = XSD + 'boolean'
export const XSD_DOUBLE = XSD + 'double'
export const XSD_INTEGER = XSD + 'integer'
export const XSD_STRING = XSD + 'string'
