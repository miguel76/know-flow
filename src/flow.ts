import { Algebra } from 'sparqlalgebrajs'
import { Table } from './table'

/**
 * Base class for flows, which are networks of know-flow operations
 * A flow take implicitly as input a sequence of RDF bindings and a knowledge
 * graph, performs some data operations which may involve querying the knowledge
 * graph, performs some data-driven actions, and finally returning some output.
 * Flows can be composed of other flows (called subflows).
 */
// eslint-disable-next-line no-unused-vars
export abstract class Flow<ReturnType> {}

export type Action<InputType, OutputType> =
  | ((input: InputType) => Promise<OutputType>)
  | ((input: InputType) => OutputType)

/**
 * Action executors are flows composed of a single (potentially async) function
 * taking as input the current sequence of bindings.
 */
export class ActionExecutor<ReturnType> extends Flow<ReturnType> {
  /** Async/sync function to be excuted */
  action: Action<Table, ReturnType>

  /**
   * Creates a new action executor
   * @param action - Async/sync function to be excuted
   */
  constructor(action: Action<Table, ReturnType>) {
    super()
    this.action = action
  }
}

/** Cascades are flows composed of a subflow and a follow up action, the latter
 * taking as input the output of the former.
 */
export class Cascade<
  SubflowReturnType,
  ActionReturnType
> extends Flow<ActionReturnType> {
  /** Subflow to be executed before the action */
  subflow: Flow<SubflowReturnType>
  /** Function to be executed as action after the subflow */
  action: Action<SubflowReturnType, ActionReturnType>

  /**
   * Creates a new cascade
   * @param subflow - Subflow to be executed before the action
   * @param action - Function to be executed as action after the subflow
   */
  constructor(
    subflow: Flow<SubflowReturnType>,
    action: Action<SubflowReturnType, ActionReturnType>
  ) {
    super()
    this.subflow = subflow
    this.action = action
  }
}

/**
 * Parallel flows are composed of an array of subflows executed in parallel.
 * The output is the array of the results of each subflow.
 */
export class Parallel<ReturnType> extends Flow<ReturnType> {
  /** Array of subflows to be executed in parallel */
  subflows: Flow<any>[]

  /**
   * Creates a new Parallel
   * @param subflows - Array of subflows to be executed in parallel
   */
  constructor(subflows: Flow<any>[]) {
    super()
    this.subflows = subflows
  }
}

/**
 * ParallelN flows are composed of an array of subflows executed in parallel,
 * having all the same return type.
 * The output is the array of the results of each subflow.
 * ParallelN, ParallelTwo and ParallelThree are subclasses of Parallel defined
 * to have more control on return types (using typescript) than using directly
 * Parallel.
 */
export class ParallelN<EachReturnType> extends Parallel<EachReturnType[]> {
  subflows: Flow<EachReturnType>[]
}

/**
 * ParallelTwo flows are composed of two subflows executed in parallel.
 * The output is the array of the results of each subflow.
 * ParallelN, ParallelTwo and ParallelThree are subclasses of Parallel defined
 * to have more control on return types (using typescript) than using directly
 */
export class ParallelTwo<ReturnType1, ReturnType2> extends Parallel<
  [ReturnType1, ReturnType2]
> {
  subflows: [Flow<ReturnType1>, Flow<ReturnType2>]
}

/**
 * ParallelThree flows are composed of three subflows executed in parallel.
 * The output is the array of the results of each subflow.
 * ParallelN, ParallelTwo and ParallelThree are subclasses of Parallel defined
 * to have more control on return types (using typescript) than using directly
 */
export class ParallelThree<
  ReturnType1,
  ReturnType2,
  ReturnType3
> extends Parallel<[ReturnType1, ReturnType2, ReturnType3]> {
  subflows: [Flow<ReturnType1>, Flow<ReturnType2>, Flow<ReturnType3>]
}

/**
 * ForEach flows consist of a subflow that is executed mutiple times (in
 * parallel), once for each input binding of either all the variables or a
 * subset of them.
 */
export class ForEach<EachReturnType> extends Flow<EachReturnType[]> {
  /** Subflow to be executed each time */
  subflow: Flow<EachReturnType>
  /** Optionally, set of variables used for the iteration. If undefined all the
   * variables in the input sequence of bindings are considered. */
  variables: string[] | undefined
  /** In the case all the variables are selected for iteration, decides if
   * multiple indentical bindings are considered in the same iteration or not. */
  distinct: boolean

  /**
   * Creates a new ForEach
   * @param subflow - Subflow to be executed each time
   * @param variablesOrDistinct - If an arrary, the set of variables used for
   * the iteration. If a boolean, all the variables in the input are
   * considered and the boolean decides if repeated bindings are taken in the
   * same iteration or not.
   */
  constructor(
    subflow: Flow<EachReturnType>,
    variablesOrDistinct: string[] | boolean
  ) {
    super()
    this.subflow = subflow
    if (Array.isArray(variablesOrDistinct)) {
      this.variables = variablesOrDistinct
      this.distinct = true
    } else {
      this.distinct = !!variablesOrDistinct
    }
  }
}

/**
 * Base class of data operations, which are the flows that are data aware.
 * Data operations manipulate in some way (depnding on the specific subclass)
 * the current sequence of bindings, without side effects.
 */
export abstract class DataOperation<ReturnType> extends Flow<ReturnType> {
  /** Subflow executed after the operation. */
  subflow: Flow<ReturnType>

  constructor(subflow: Flow<ReturnType>) {
    super()
    this.subflow = subflow
  }
}

/**
 * Let operations replace the value of a variable with the value taken from
 * another variable.
 * It optionally hides the variable originally holding the value
 */
export class Let<ReturnType> extends DataOperation<ReturnType> {
  /** Name of the variable whose value is used. */
  currVarname: string
  /** Name of the variable to which the value is assigned. */
  newVarname: string
  /** If true, the original variable (`currVarname`) is hidden.  */
  hideCurrVar: boolean

  /**
   * Creates a new Let
   * @param subflow - Subflow executed after the operation.
   * @param currVarname - Name of the variable whose value is used.
   * @param newVarname - Name of the variable to which the value is assigned.
   * @param hideCurrVar - If true, the original variable (`currVarname`) is
   * hidden.
   */
  constructor(
    subflow: Flow<ReturnType>,
    currVarname: string,
    newVarname: string,
    hideCurrVar: boolean
  ) {
    super(subflow)
    this.currVarname = currVarname
    this.newVarname = newVarname
    this.hideCurrVar = hideCurrVar
  }
}

/**
 * Object to configure the renaming of a variable
 */
export type SingleVarRenameConfig = {
  /** Current of the variable. */
  currVarname: string
  /** New name of the variable. */
  newVarname: string
  /** If true, the original variable (`currVarname`) is hidden.  */
  hideCurrVar: boolean
}

/**
 * Rename operations rename a set of variables
 */
export class Rename<ReturnType> extends DataOperation<ReturnType> {
  /** Name of the variable whose value is used. */
  renamings: SingleVarRenameConfig[]

  /**
   * Creates a new Let
   * @param subflow - Subflow executed after the operation.
   * @param renamings - Configuration of the renamings applied.
   */
  constructor(subflow: Flow<ReturnType>, renamings: SingleVarRenameConfig[]) {
    super(subflow)
    this.renamings = renamings
  }
}

/**
 * Hide operations omit a set of variables from the bindings
 */
export class Hide<ReturnType> extends DataOperation<ReturnType> {
  /** Names of the variables to omit. */
  variables: string[]

  /**
   * Creates a new Hide
   * @param subflow - Subflow executed after the operation.
   * @param variables - Names of the variables to omit.
   */
  constructor(subflow: Flow<ReturnType>, variables: string[]) {
    super(subflow)
    this.variables = variables
  }
}

/**
 * Join operations perform a Join between the current sequence of bindings and
 * the output of a SPARQL subquery.
 * @see {@link https://www.w3.org/TR/sparql11-query/#defn_algJoin}
 */
export class Join<ReturnType> extends DataOperation<ReturnType> {
  /** SPARQL subquery which is joined with the current sequence of bindings. */
  right: Algebra.Operation

  /**
   * Creates a new Join
   * @param subflow - Subflow executed after the operation.
   * @param right - SPARQL subquery which is joined with the current sequence of
   * bindings.
   */
  constructor(subflow: Flow<ReturnType>, right: Algebra.Operation) {
    super(subflow)
    this.right = right
  }
}

/**
 * Filter operations perform a Filter over the current sequence of bindings.
 * @see {@link https://www.w3.org/TR/sparql11-query/#defn_algFilter}
 */
export class Filter<ReturnType> extends DataOperation<ReturnType> {
  /** Expression used for filter the current sequence of bindings. */
  expression: Algebra.Expression

  /**
   * Creates a new Filter
   * @param subflow - Subflow executed after the operation.
   * @param expression - Expression used for filter the current sequence of
   * bindings.
   */
  constructor(subflow: Flow<ReturnType>, expression: Algebra.Expression) {
    super(subflow)
    this.expression = expression
  }
}

export class Aggregate<ReturnType> extends DataOperation<ReturnType> {
  aggregates: Algebra.BoundAggregate[]

  constructor(subflow: Flow<ReturnType>, aggregates: Algebra.BoundAggregate[]) {
    super(subflow)
    this.aggregates = aggregates
  }
}

export class Slice<ReturnType> extends DataOperation<ReturnType> {
  start: number
  length?: number

  constructor(subflow: Flow<ReturnType>, start: number, length?: number) {
    super(subflow)
    this.start = start
    this.length = length
  }
}
