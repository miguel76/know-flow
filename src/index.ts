import * as Tasks from './task';
import TaskFactory from './taskFactory';
import TaskBuilder from './taskBuilder';
import {stringifyTask} from './utils';
import TaskEngine from './taskEngine';

export {
    Tasks,
    TaskFactory, stringifyTask, TaskEngine,
    TaskBuilder
};


