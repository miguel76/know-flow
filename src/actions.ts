import * as RDF from 'rdf-js'
import { Table, TableSync, syncTable } from './table'
import { Action, Flow } from './flow'
import { Bindings } from '@comunica/types'
import { Map } from 'immutable'
import { DEFAULT_INPUT_VARNAME } from './constants'

type ValueOf<T> = T[keyof T];

/**
 * Creates an Action that returns a constant value.
 * @param value - Value to be returned.
 * @returns The new Action.
 */
export function constant<ReturnType>(
  value: ReturnType
): Action<void, ReturnType> {
  return () => value
}

/**
 * Creates an Action that calls an inner action on the entire input sequence.
 * @param action - Action executed on the whole input sequence.
 * @returns The new Action.
 */
export function onAll<ReturnType>(
  fun: Action<TableSync, ReturnType>
): Action<Table, ReturnType> {
  return async (table: Table) => fun(await syncTable(table))
}

/**
 * Creates an Action that calls an inner action on the entire input sequence.
 * @param action - Action executed on the whole input sequence.
 * @returns The new Action.
 */
 export function onThisVariables<ReturnType, VarnameSet extends string>(
  varnames: VarnameSet[],
  fun: Action<Map<VarnameSet,RDF.Term>, ReturnType>
): Flow<ReturnType> {

  return async (table: Table) => fun(await syncTable(table))
}

/**
 * Creates an Action that calls an inner action on the first binding
 * of the sequence, ignoring the rest.
 * @param action - Action executed on the first binding.
 * @returns The new Action.
 */
export function onFirst<ReturnType>(
  action: Action<Bindings, ReturnType>,
  acceptEmpty: boolean = true
): Action<Table, ReturnType> {
  return async (table) => {
    const oneTuple = await syncTable(table, 1)
    let bindings: Bindings
    if (oneTuple.bindingsArray.length === 0) {
      if (acceptEmpty) {
        bindings = Map<string, RDF.Term>({})
      } else {
        throw new Error('Expected at least a value, zero found')
      }
    } else {
      bindings = Map<string, RDF.Term>(oneTuple.bindingsArray[0])
    }
    return action(bindings)
  }
}

/**
 * Creates an Action that calls an inner action on the value of the default
 * variable in  the first binding of the sequence, ignoring the rest.
 * @param action - Action executed on the value.
 * @returns The new Action.
 */
export function onFirstDefault<ReturnType>(
  action: Action<RDF.Term, ReturnType>
): Action<Table, ReturnType> {
  return onFirst(async (bindings: Bindings) => action(bindings.get(DEFAULT_INPUT_VARNAME)))
}
