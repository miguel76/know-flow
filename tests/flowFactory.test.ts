import { FlowFactory, FlowEngine, Actions } from '../src/index'

import { newEngine } from '@comunica/actor-init-sparql-file'
const path = require('path')

const ff = new FlowFactory({
  prefixes: {
    ex: 'http://example.org/',
    rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
    rdfs: 'http://www.w3.org/2000/01/rdf-schema#'
  }
})

// let engine = newEngine();
// let proxyEngine: IQueryEngine = {
//     query: async (query: string | Algebra.Operation, queryContext: any) => {
//         console.log('');
//         console.log('Executing...');
//         console.log(typeof query === 'string' ? query : toSparqlFragment(query));
//         const res = <IActorQueryOperationOutputBindings>await engine.query(query, queryContext);
//         console.log('Result variables :' + res.variables);
//         return res;
//     },
//     getResultMediaTypes: function (context?: ActionContext): Promise<Record<string, number>> {
//         throw new Error('Function not implemented.');
//     },
//     getResultMediaTypeFormats: function (context?: ActionContext): Promise<Record<string, string>> {
//         throw new Error('Function not implemented.');
//     },
//     resultToString: function (queryResult: IActorQueryOperationOutput, mediaType?: string, context?: any) {
//         throw new Error('Function not implemented.');
//     },
//     invalidateHttpCache: function (url?: string): Promise<any> {
//         throw new Error('Function not implemented.');
//     }
// };

const fe = new FlowEngine({
  engine: newEngine(),
  queryContext: {
    sources: [path.join(__dirname, '../tests/test-data.ttl')]
  }
})

const showBindings = ff.createActionExecutor(Actions.onAll((b) => b))
const showOneBinding = ff.createActionExecutor(Actions.onFirst((b) => b))
const showOne = ff.createActionExecutor(Actions.onFirstDefault((b) => b))

const flows = {
  constant: ff.createActionExecutor(Actions.constant(42)),
  'action constant': ff.createActionExecutor(() => {
    return 42
  }),
  'action show bindings': showBindings,
  'action show one set of bindings': showOneBinding,
  'action show default bindings': showOne,
  'undefined term reader': ff.createTermReader(),
  'undefined value reader': ff.createValueReader(),
  'single default binding show value': ff.createValues({
    bindings: 'ex:Res1',
    subflow: ff.createValueReader()
  }),
  'single default binding show term': ff.createValues({
    bindings: 'ex:Res1',
    subflow: ff.createTermReader()
  }),
  'single default binding show all': ff.createValues({
    bindings: 'ex:Res1',
    subflow: showBindings
  }),
  traversal: ff.createValues({
    bindings: 'ex:Res1',
    subflow: ff.createValueReader({ path: 'ex:prop1' })
  }),
  'traversal path': ff.createValues({
    bindings: 'ex:Res1',
    subflow: ff.createValueReader({ path: 'ex:prop1/ex:prop2' })
  }),
  'multiple default bindings': ff.createValues({
    bindings: ['ex:Res1', 'ex:Res2', 'ex:Res3'],
    subflow: showBindings
  }),
  join: ff.createJoin({
    right: '?s ?p ?o',
    subflow: showBindings
  }),
  'empty parallel': ff.createParallel([]),
  parallel: ff.createValues({
    bindings: ['ex:Res1', 'ex:Res2', 'ex:Res3'],
    subflow: ff.createParallel([
      ff.createValueReader(),
      ff.createValueReader({ path: 'ex:prop1' }),
      ff.createValueReader({ path: 'ex:prop2' }),
      ff.createValueReader({ path: 'ex:prop3' })
    ])
  }),
  'foreach value reader': ff.createValues({
    bindings: ['ex:Res1', 'ex:Res2', 'ex:Res3', '"pippo"', '42', '3.14'],
    subflow: ff.createForEach(ff.createValueReader())
  }),
  'foreach traversal value reader': ff.createValues({
    bindings: ['ex:Res1', 'ex:Res2', 'ex:Res3', '"pippo"', '42', '3.14'],
    subflow: ff.createForEach({
      select: { path: 'ex:prop1' },
      subflow: ff.createValueReader()
    })
  }),
  'foreach foreach string reader': ff.createValues({
    bindings: ['ex:Res1', 'ex:Res2', 'ex:Res3'],
    subflow: ff.createForEach(
      ff.createForEach({
        select: { path: 'ex:prop1' },
        subflow: ff.createStringReader()
      })
    )
  }),
  'foreach foreach value reader': ff.createValues({
    bindings: ['ex:Res1', 'ex:Res2', 'ex:Res3'],
    subflow: ff.createForEach(
      ff.createForEach({
        select: { path: 'ex:prop1' },
        subflow: ff.createValueReader()
      })
    )
  }),
  'all triples foreach x 3': ff.createJoin({
    right: '?s ?p ?o',
    subflow: ff.createForEach({
      select: ['?s'],
      subflow: ff.createForEach({
        select: ['?p'],
        subflow: ff.createForEach({
          select: ['?o'],
          subflow: ff.createParallelDict({
            s: ff.createValueReader({ var: '?s' }),
            p: ff.createValueReader({ var: '?p' }),
            o: ff.createValueReader({ var: '?o' })
          })
        })
      })
    })
  }),
  'all triples foreach x 2': ff.createJoin({
    right: '?s ?p ?o',
    subflow: ff.createForEach({
      select: ['?s'],
      subflow: ff.createForEach({
        select: ['?p', '?o'],
        subflow: ff.createParallelDict({
          s: ff.createValueReader({ var: '?s' }),
          p: ff.createValueReader({ var: '?p' }),
          o: ff.createValueReader({ var: '?o' })
        })
      })
    })
  }),
  'all triples foreach x 1': ff.createJoin({
    right: '?s ?p ?o',
    subflow: ff.createForEach({
      select: ['?s', '?p', '?o'],
      subflow: ff.createParallelDict({
        s: ff.createValueReader({ var: '?s' }),
        p: ff.createValueReader({ var: '?p' }),
        o: ff.createValueReader({ var: '?o' })
      })
    })
  }),
  'all triples foreach x 1 implicit': ff.createJoin({
    right: '?s ?p ?o',
    subflow: ff.createForEach({
      select: { allVars: true },
      subflow: ff.createParallelDict({
        s: ff.createValueReader({ var: '?s' }),
        p: ff.createValueReader({ var: '?p' }),
        o: ff.createValueReader({ var: '?o' })
      })
    })
  })
}

// jest.setTimeout(60000);

Object.entries(flows).forEach(([label, flow]) => {
  test('run ' + label, () => expect(fe.run(flow)).resolves.toMatchSnapshot())
})

// Object.entries(flows).forEach(([label, flow]) => {
//     test('run ' + label, async () => {
//         console.log(flow);
//         let result = await te.run(flow);
//         console.log(result);
//         expect(result).toMatchSnapshot();
//     });
// });

// Object.entries(flows).forEach(([label, flow]) => {
//     test('run ' + label, done => {
//         console.log(flow);
//         te.run(flow).then(result => {
//             expect(result).toMatchSnapshot();
//             done();
//         });
//     });
// });
