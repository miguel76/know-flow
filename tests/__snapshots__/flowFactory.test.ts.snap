// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run action constant 1`] = `42`;

exports[`run action show bindings 1`] = `
Object {
  "bindingsArray": Array [
    Object {},
  ],
  "canContainUndefs": false,
  "variables": Array [],
}
`;

exports[`run action show default bindings 1`] = `undefined`;

exports[`run action show one set of bindings 1`] = `Immutable.Map {}`;

exports[`run all triples foreach x 1 1`] = `
Array [
  Object {
    "o": "42",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "value1",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "http://example.org/Res2",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "http://example.org/Res3",
    "p": "http://example.org/prop2",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "http://example.org/Res1",
    "p": "http://example.org/prop3",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "3.14",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res2",
  },
  Object {
    "o": "http://example.org/Res1",
    "p": "http://example.org/prop2",
    "s": "http://example.org/Res2",
  },
]
`;

exports[`run all triples foreach x 1 implicit 1`] = `
Array [
  Object {
    "o": "value1",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "42",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "http://example.org/Res2",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "http://example.org/Res3",
    "p": "http://example.org/prop2",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "http://example.org/Res1",
    "p": "http://example.org/prop3",
    "s": "http://example.org/Res1",
  },
  Object {
    "o": "3.14",
    "p": "http://example.org/prop1",
    "s": "http://example.org/Res2",
  },
  Object {
    "o": "http://example.org/Res1",
    "p": "http://example.org/prop2",
    "s": "http://example.org/Res2",
  },
]
`;

exports[`run all triples foreach x 2 1`] = `
Array [
  Array [
    Object {
      "o": "42",
      "p": "http://example.org/prop1",
      "s": "http://example.org/Res1",
    },
    Object {
      "o": "value1",
      "p": "http://example.org/prop1",
      "s": "http://example.org/Res1",
    },
    Object {
      "o": "http://example.org/Res2",
      "p": "http://example.org/prop1",
      "s": "http://example.org/Res1",
    },
    Object {
      "o": "http://example.org/Res3",
      "p": "http://example.org/prop2",
      "s": "http://example.org/Res1",
    },
    Object {
      "o": "http://example.org/Res1",
      "p": "http://example.org/prop3",
      "s": "http://example.org/Res1",
    },
  ],
  Array [
    Object {
      "o": "3.14",
      "p": "http://example.org/prop1",
      "s": "http://example.org/Res2",
    },
    Object {
      "o": "http://example.org/Res1",
      "p": "http://example.org/prop2",
      "s": "http://example.org/Res2",
    },
  ],
]
`;

exports[`run all triples foreach x 3 1`] = `
Array [
  Array [
    Array [
      Object {
        "o": "42",
        "p": "http://example.org/prop1",
        "s": "http://example.org/Res1",
      },
      Object {
        "o": "value1",
        "p": "http://example.org/prop1",
        "s": "http://example.org/Res1",
      },
      Object {
        "o": "http://example.org/Res2",
        "p": "http://example.org/prop1",
        "s": "http://example.org/Res1",
      },
    ],
    Array [
      Object {
        "o": "http://example.org/Res3",
        "p": "http://example.org/prop2",
        "s": "http://example.org/Res1",
      },
    ],
    Array [
      Object {
        "o": "http://example.org/Res1",
        "p": "http://example.org/prop3",
        "s": "http://example.org/Res1",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "o": "3.14",
        "p": "http://example.org/prop1",
        "s": "http://example.org/Res2",
      },
    ],
    Array [
      Object {
        "o": "http://example.org/Res1",
        "p": "http://example.org/prop2",
        "s": "http://example.org/Res2",
      },
    ],
  ],
]
`;

exports[`run constant 1`] = `42`;

exports[`run empty parallel 1`] = `Array []`;

exports[`run foreach foreach string reader 1`] = `
Array [
  Array [
    "42",
    "value1",
    "http://example.org/Res2",
  ],
  Array [
    "3.14",
  ],
  Array [],
]
`;

exports[`run foreach foreach value reader 1`] = `
Array [
  Array [
    "42",
    "value1",
    "http://example.org/Res2",
  ],
  Array [
    "3.14",
  ],
  Array [],
]
`;

exports[`run foreach traversal value reader 1`] = `
Array [
  "3.14",
  "42",
  "value1",
  "http://example.org/Res2",
]
`;

exports[`run foreach value reader 1`] = `
Array [
  "3.14",
  "42",
  "pippo",
  "http://example.org/Res1",
  "http://example.org/Res2",
  "http://example.org/Res3",
]
`;

exports[`run join 1`] = `
Object {
  "bindingsArray": Array [
    Object {
      "?o": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#string",
        },
        "language": "",
        "termType": "Literal",
        "value": "value1",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop1",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
    Object {
      "?o": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#integer",
        },
        "language": "",
        "termType": "Literal",
        "value": "42",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop1",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
    Object {
      "?o": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res2",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop1",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
    Object {
      "?o": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res3",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop2",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
    Object {
      "?o": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop3",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
    Object {
      "?o": Object {
        "datatype": Object {
          "termType": "NamedNode",
          "value": "http://www.w3.org/2001/XMLSchema#decimal",
        },
        "language": "",
        "termType": "Literal",
        "value": "3.14",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop1",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res2",
      },
    },
    Object {
      "?o": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
      "?p": Object {
        "termType": "NamedNode",
        "value": "http://example.org/prop2",
      },
      "?s": Object {
        "termType": "NamedNode",
        "value": "http://example.org/Res2",
      },
    },
  ],
  "canContainUndefs": false,
  "variables": Array [
    "?s",
    "?p",
    "?o",
  ],
}
`;

exports[`run multiple default bindings 1`] = `
Object {
  "bindingsArray": Array [
    Object {
      "?_": NamedNode {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
    Object {
      "?_": NamedNode {
        "termType": "NamedNode",
        "value": "http://example.org/Res2",
      },
    },
    Object {
      "?_": NamedNode {
        "termType": "NamedNode",
        "value": "http://example.org/Res3",
      },
    },
  ],
  "canContainUndefs": false,
  "variables": Array [
    "?_",
  ],
}
`;

exports[`run parallel 1`] = `
Array [
  "http://example.org/Res1",
  "value1",
  "http://example.org/Res1",
  "http://example.org/Res1",
]
`;

exports[`run single default binding show all 1`] = `
Object {
  "bindingsArray": Array [
    Object {
      "?_": NamedNode {
        "termType": "NamedNode",
        "value": "http://example.org/Res1",
      },
    },
  ],
  "canContainUndefs": false,
  "variables": Array [
    "?_",
  ],
}
`;

exports[`run single default binding show term 1`] = `
NamedNode {
  "termType": "NamedNode",
  "value": "http://example.org/Res1",
}
`;

exports[`run single default binding show value 1`] = `"http://example.org/Res1"`;

exports[`run traversal 1`] = `"value1"`;

exports[`run traversal path 1`] = `"http://example.org/Res1"`;

exports[`run undefined term reader 1`] = `undefined`;

exports[`run undefined value reader 1`] = `null`;
